### PR TITLE
feat: add configuration classes

### PR DESCRIPTION
**Pull request by pgautier404**
This commit updates the Python SDK to use the [standard configuration classes](https://github.com/momentohq/standards-and-practices/blob/main/docs/client-specifications/configuration.md) and the credential provider interface. It also updates ttl and client deadline values so they are expressed as `timedelta` objects instead of integers. Finally, unit tests that made assertions about exception messages have also been updated, as they were silently ignoring the message text.

### CODE REVIEW

File: src/momento/_utilities/_data_validation.py [hunk 1]
```diff
@@ -21,18 +22,18 @@ def _as_bytes(
...
+    if not isinstance(ttl, timedelta) or ttl.total_seconds() < 0:
+        raise errors.InvalidArgumentError("TTL timedelta must be a non-negative integer")
 
 
-def _validate_ttl_minutes(ttl_minutes: int) -> None:
-    if not isinstance(ttl_minutes, int) or ttl_minutes < 0:
-        raise errors.InvalidArgumentError("TTL Minutes must be a non-negative integer")
+def _validate_ttl_seconds(ttl_seconds: int) -> None:
+    if not isinstance(ttl_seconds, int) or ttl_seconds < 0:
+        raise errors.InvalidArgumentError("TTL Seconds must be a non-negative integer")
```
Comment (schwern) [inline src/momento/_utilities/_data_validation.py:32]: I think this is only used internally because prepare_set_request takes seconds. If prepare_set_request is changed to use timedelta this can be deleted.

File: src/momento/aio/_scs_data_client.py [hunk 2]
```diff
@@ -35,21 +38,17 @@ def _make_metadata(cache_name: str) -> Metadata:
...
-        operation_timeout_ms: Optional[int],
-    ):
+    def __init__(self, configuration: Configuration, credential_provider: CredentialProvider, default_ttl: timedelta):
+        endpoint = credential_provider.get_cache_endpoint()
         self._logger = logs.logger
         self._logger.debug("Simple cache data client instantiated with endpoint: %s", endpoint)
-        self._default_deadline_seconds = (
-            _DEFAULT_DEADLINE_SECONDS if not operation_timeout_ms else operation_timeout_ms / 1000.0
+        default_deadline: timedelta = (
+            configuration.get_transport_strategy().get_grpc_configuration().get_deadline() or _DEFAULT_DEADLINE_SECONDS
```
Comment (schwern) [inline src/momento/aio/_scs_data_client.py:46]: This suggests a wrapper method to hide all that and to provide a default.

File: src/momento/aio/_scs_data_client.py [hunk 3]
```diff
@@ -35,21 +38,17 @@ def _make_metadata(cache_name: str) -> Metadata:
...
         )
-        self._grpc_manager = _scs_grpc_manager._DataGrpcManager(auth_token, endpoint)
-        _validate_ttl(default_ttl_seconds)
-        self._default_ttlSeconds = default_ttl_seconds
+        self._default_deadline_seconds = int(default_deadline.total_seconds())
+        self._grpc_manager = _scs_grpc_manager._DataGrpcManager(credential_provider)
+        _validate_ttl(default_ttl)
+        self._default_ttlSeconds = int(default_ttl.total_seconds())
```
Comment (schwern) [inline src/momento/aio/_scs_data_client.py:51]: Only used by `prepare_set_request`.

File: src/momento/aio/_scs_data_client.py [hunk 4]
```diff
@@ -69,6 +68,7 @@ async def execute_set_request_fn(req: _SetRequest) -> _SetResponse:
                 timeout=self._default_deadline_seconds,
             )
 
+        ttl_seconds: Optional[int] = int(ttl.total_seconds()) if ttl else None
```
Comment (schwern) [inline src/momento/aio/_scs_data_client.py:71]: Only used by `prepare_set_request`.

File: src/momento/aio/simple_cache_client.py [hunk 5]
```diff
@@ -205,15 +194,15 @@ async def set(
...
+        ttl: Optional[timedelta] = None,
     ) -> CacheSetResponse:
         """Stores an item in cache
 
         Args:
             cache_name: Name of the cache to store the item in.
             key (string or bytes): The key to be used to store item.
             value (string or bytes): The value to be stored.
-            ttl_seconds (Optional): Time to live in cache in seconds. If not provided, then default TTL for the cache
+            ttl (Optional): Time to live in cache timedelta. If not provided, then default TTL for the cache
```
Comment (schwern) [inline src/momento/aio/simple_cache_client.py:205]:
```suggestion
            ttl (Optional[timedelta]): Time to live in cache timedelta. If not provided, then default TTL for the cache
```

File: src/momento/internal/synchronous/_scs_data_client.py [hunk 6]
```diff
@@ -36,32 +39,35 @@ class _ScsDataClient:
...
-        default_ttl_seconds: int,
-        request_timeout_ms: Optional[int],
+        configuration: Configuration,
+        credential_provider: CredentialProvider,
+        default_ttl: timedelta,
     ):
-        self._default_deadline_seconds = (
-            _DEFAULT_DEADLINE_SECONDS if not request_timeout_ms else request_timeout_ms / 1000.0
+        default_deadline: timedelta = (
+            configuration.get_transport_strategy().get_grpc_configuration().get_deadline() or _DEFAULT_DEADLINE_SECONDS
```
Comment (schwern) [inline src/momento/internal/synchronous/_scs_data_client.py:47]: A wrapper method would help here.

File: src/momento/internal/synchronous/_scs_data_client.py [hunk 7]
```diff
@@ -36,32 +39,35 @@ class _ScsDataClient:
...
         )
-        self._grpc_manager = _scs_grpc_manager._DataGrpcManager(auth_token, endpoint)
-        _validate_ttl(default_ttl_seconds)
-        self._default_ttlSeconds = default_ttl_seconds
+        self._default_deadline_seconds = int(default_deadline.total_seconds())
+        self._grpc_manager = _scs_grpc_manager._DataGrpcManager(
+            credential_provider.get_auth_token(), credential_provider.get_cache_endpoint()
+        )
+        _validate_ttl(default_ttl)
+        self._default_ttlSeconds = int(default_ttl.total_seconds())
```
Comment (schwern) [inline src/momento/internal/synchronous/_scs_data_client.py:54]: Only used by prepare_set_request.

File: src/momento/internal/synchronous/_scs_data_client.py [hunk 8]
```diff
@@ -36,32 +39,35 @@ class _ScsDataClient:
...
     ) -> cache_operation_types.CacheSetResponse:
         def execute_set_request_fn(req: _SetRequest) -> _SetResponse:
-            return self._getStub().Set(
+            return self._get_stub().Set(
                 req,
                 metadata=_make_metadata(cache_name),
                 timeout=self._default_deadline_seconds,
             )
 
+        ttl_seconds: Optional[int] = int(ttl.total_seconds()) if ttl else None
```
Comment (schwern) [inline src/momento/internal/synchronous/_scs_data_client.py:70]: Only used by prepare_set_request

File: src/momento/simple_cache_client.py [hunk 9]
```diff
@@ -11,38 +12,35 @@
...
+            default_ttl (timedelta): A default Time To Live timedelta for cache objects created by this client.
+                It is possible to override this setting when calling the set method.
         Raises:
             IllegalArgumentError: If method arguments fail validations.
         """
-        _validate_request_timeout(request_timeout_ms)
-
-        endpoints = _momento_endpoint_resolver.resolve(auth_token)
-        self._control_client = _ScsControlClient(auth_token, endpoints.control_endpoint)
+        _validate_request_timeout(configuration.get_transport_strategy().get_grpc_configuration().get_deadline())
```
Comment (schwern) [inline src/momento/simple_cache_client.py:37]: Should this validation happen in the GRPC config object?

File: tests/test_momento.py [hunk 10]
```diff
@@ -30,29 +34,32 @@ def test_create_cache_get_set_values_and_delete_cache(client: SimpleCacheClient,
...
-def test_init_throws_exception_when_client_uses_negative_request_timeout_ms(auth_token: str, default_ttl_seconds: int):
-    with pytest.raises(errors.InvalidArgumentError) as cm:
-        SimpleCacheClient(auth_token, default_ttl_seconds, -1)
-        assert cm.exception == "Request timeout must be greater than zero."
+def test_init_throws_exception_when_client_uses_negative_request_timeout_ms(
+    configuration: Configuration, credential_provider: CredentialProvider, default_ttl_seconds: int
+):
+    with pytest.raises(errors.InvalidArgumentError, match="Request timeout must be greater than zero."):
+        configuration = configuration.with_client_timeout(-1)
+        SimpleCacheClient(configuration, credential_provider, default_ttl_seconds)
```
Comment (schwern) [inline tests/test_momento.py:54]: Doesn't this take a timedelta, not an int?

How did this pass tests? Does an int get cast to a timedelta?

File: tests/test_momento.py [hunk 11]
```diff
@@ -30,29 +34,32 @@ def test_create_cache_get_set_values_and_delete_cache(client: SimpleCacheClient,
...
-def test_init_throws_exception_when_client_uses_zero_request_timeout_ms(auth_token: str, default_ttl_seconds: int):
-    with pytest.raises(errors.InvalidArgumentError) as cm:
-        SimpleCacheClient(auth_token, default_ttl_seconds, 0)
-        assert cm.exception == "Request timeout must be greater than zero."
+def test_init_throws_exception_when_client_uses_zero_request_timeout_ms(
+    configuration: Configuration, credential_provider: CredentialProvider, default_ttl_seconds: int
+):
+    with pytest.raises(errors.InvalidArgumentError, match="Request timeout must be greater than zero."):
+        configuration = configuration.with_client_timeout(0)
+        SimpleCacheClient(configuration, credential_provider, default_ttl_seconds)
```
Comment (schwern) [inline tests/test_momento.py:62]: Same.

File: tests/test_momento.py [hunk 12]
```diff
@@ -71,21 +78,21 @@ def test_create_cache_throws_exception_for_empty_cache_name(
...
         client.create_cache(1)
-        assert cm.exception == "Cache name must be a non-empty string"
 
 
-def test_create_cache_throws_authentication_exception_for_bad_token(bad_auth_token: str, default_ttl_seconds: int):
-    with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client:
+def test_create_cache_throws_authentication_exception_for_bad_token(
+    configuration: Configuration, bad_token_credential_provider: EnvMomentoTokenProvider, default_ttl_seconds: int
+):
+    with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:95]: same

File: tests/test_momento.py [hunk 13]
```diff
@@ -124,13 +131,14 @@ def test_delete_cache_throws_exception_for_empty_cache_name(
...
         client.delete_cache(1)
-        assert cm.exception == "Cache name must be a non-empty string"
 
 
-def test_delete_cache_throws_authentication_exception_for_bad_token(bad_auth_token: str, default_ttl_seconds: int):
-    with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client:
+def test_delete_cache_throws_authentication_exception_for_bad_token(
+    configuration: Configuration, bad_token_credential_provider: EnvMomentoTokenProvider, default_ttl_seconds: int
+):
+    with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:141]: same

File: tests/test_momento.py [hunk 14]
```diff
@@ -155,8 +163,10 @@ def test_list_caches_succeeds(client: SimpleCacheClient, cache_name: str):
         client.delete_cache(cache_name)
 
 
-def test_list_caches_throws_authentication_exception_for_bad_token(bad_auth_token: str, default_ttl_seconds: int):
-    with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client:
+def test_list_caches_throws_authentication_exception_for_bad_token(
+    configuration: Configuration, bad_token_credential_provider: EnvMomentoTokenProvider, default_ttl_seconds: int
+):
+    with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:169]: same

File: tests/test_momento.py [hunk 15]
```diff
@@ -272,41 +280,43 @@ def test_set_with_null_value_throws_exception(client: SimpleCacheClient, cache_n
...
 
 def test_set_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, cache_name: str, default_ttl_seconds: int
+    configuration: Configuration,
+    bad_token_credential_provider: EnvMomentoTokenProvider,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client:
+    with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:310]: same

File: tests/test_momento.py [hunk 16]
```diff
@@ -272,41 +280,43 @@ def test_set_with_null_value_throws_exception(client: SimpleCacheClient, cache_n
...
             client.set(cache_name, "foo", "bar")
 
 
-def test_set_throws_timeout_error_for_short_request_timeout(auth_token: str, cache_name: str, default_ttl_seconds: int):
-    with SimpleCacheClient(auth_token, default_ttl_seconds, request_timeout_ms=1) as client:
+def test_set_throws_timeout_error_for_short_request_timeout(
+    configuration: Configuration, credential_provider: CredentialProvider, cache_name: str, default_ttl_seconds: int
+):
+    configuration = configuration.with_client_timeout(timedelta(milliseconds=1))
+    with SimpleCacheClient(configuration, credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:319]: same

File: tests/test_momento.py [hunk 17]
```diff
@@ -344,27 +352,31 @@ def test_get_with_null_key_throws_exception(client: SimpleCacheClient, cache_nam
...
 
 def test_get_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, cache_name: str, default_ttl_seconds: int
+    configuration: Configuration,
+    bad_token_credential_provider: EnvMomentoTokenProvider,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client:
+    with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:370]: same

File: tests/test_momento.py [hunk 18]
```diff
@@ -344,27 +352,31 @@ def test_get_with_null_key_throws_exception(client: SimpleCacheClient, cache_nam
...
             client.get(cache_name, "foo")
 
 
-def test_get_throws_timeout_error_for_short_request_timeout(auth_token: str, cache_name: str, default_ttl_seconds: int):
-    with SimpleCacheClient(auth_token, default_ttl_seconds, request_timeout_ms=1) as client:
+def test_get_throws_timeout_error_for_short_request_timeout(
+    configuration: Configuration, credential_provider: CredentialProvider, cache_name: str, default_ttl_seconds: int
+):
+    configuration = configuration.with_client_timeout(timedelta(milliseconds=1))
+    with SimpleCacheClient(configuration, credential_provider, default_ttl_seconds) as client:
```
Comment (schwern) [inline tests/test_momento.py:379]: same

File: tests/test_momento_async.py [hunk 19]
```diff
@@ -30,33 +34,33 @@ async def test_create_cache_get_set_values_and_delete_cache(client_async: Simple
...
 
-async def test_init_throws_exception_for_non_jwt_token(default_ttl_seconds: int):
-    with pytest.raises(errors.InvalidArgumentError) as cm:
-        SimpleCacheClient("notanauthtoken", default_ttl_seconds)
-        assert cm.exception == "Invalid Auth token."
+async def test_init_throws_exception_for_non_jwt_token(configuration: Configuration, default_ttl_seconds: int):
+    with pytest.raises(errors.InvalidArgumentError, match="Invalid Auth token.") as cm:
+        os.environ["BAD_AUTH_TOKEN"] = "notanauthtoken"
+        credential_provider = EnvMomentoTokenProvider("BAD_AUTH_TOKEN")
+        SimpleCacheClient(configuration, credential_provider, default_ttl_seconds)
```
Comment (schwern) [inline tests/test_momento_async.py:47]: same

File: tests/test_momento_async.py [hunk 20]
```diff
@@ -30,33 +34,33 @@ async def test_create_cache_get_set_values_and_delete_cache(client_async: Simple
...
 async def test_init_throws_exception_when_client_uses_negative_request_timeout_ms(
-    auth_token: str, default_ttl_seconds: int
+    configuration: Configuration, credential_provider: CredentialProvider, default_ttl_seconds: int
 ):
-    with pytest.raises(errors.InvalidArgumentError) as cm:
-        SimpleCacheClient(auth_token, default_ttl_seconds, -1)
-        assert cm.exception == "Request timeout must be greater than zero."
+    with pytest.raises(errors.InvalidArgumentError, match="Request timeout must be greater than zero."):
+        configuration = configuration.with_client_timeout(timedelta(seconds=-1))
+        SimpleCacheClient(configuration, credential_provider, default_ttl_seconds)
```
Comment (schwern) [inline tests/test_momento_async.py:55]: same

File: tests/test_momento_async.py [hunk 21]
```diff
@@ -30,33 +34,33 @@ async def test_create_cache_get_set_values_and_delete_cache(client_async: Simple
...
 async def test_init_throws_exception_when_client_uses_zero_request_timeout_ms(
-    auth_token: str, default_ttl_seconds: int
+    configuration: Configuration, credential_provider: CredentialProvider, default_ttl_seconds: int
 ):
-    with pytest.raises(errors.InvalidArgumentError) as cm:
-        SimpleCacheClient(auth_token, default_ttl_seconds, 0)
-        assert cm.exception == "Request timeout must be greater than zero."
+    with pytest.raises(errors.InvalidArgumentError, match="Request timeout must be greater than zero."):
+        configuration = configuration.with_client_timeout(timedelta(seconds=0))
+        SimpleCacheClient(configuration, credential_provider, default_ttl_seconds)
```
Comment (schwern) [inline tests/test_momento_async.py:63]: same

File: tests/test_momento_async.py [hunk 22]
```diff
@@ -77,23 +81,21 @@ async def test_create_cache_throws_exception_for_empty_cache_name(
...
         await client_async.create_cache(1)
-        assert cm.exception == "Cache name must be a non-empty string"
 
 
 async def test_create_cache_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, default_ttl_seconds: int
+    bad_token_credential_provider: EnvMomentoTokenProvider, configuration: Configuration, default_ttl_seconds: int
 ):
-    async with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client_async:
+    async with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:98]: same

File: tests/test_momento_async.py [hunk 23]
```diff
@@ -132,15 +134,14 @@ async def test_delete_cache_throws_exception_for_empty_cache_name(
...
         await client_async.delete_cache(1)
-        assert cm.exception == "Cache name must be a non-empty string"
 
 
 async def test_delete_cache_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, default_ttl_seconds: int
+    bad_token_credential_provider: EnvMomentoTokenProvider, configuration: Configuration, default_ttl_seconds: int
 ):
-    async with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client_async:
+    async with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:144]: same

File: tests/test_momento_async.py [hunk 24]
```diff
@@ -165,8 +166,10 @@ async def test_list_caches_succeeds(client_async: SimpleCacheClient, cache_name:
         await client_async.delete_cache(cache_name)
 
 
-async def test_list_caches_throws_authentication_exception_for_bad_token(bad_auth_token: str, default_ttl_seconds: int):
-    async with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client_async:
+async def test_list_caches_throws_authentication_exception_for_bad_token(
+    bad_token_credential_provider: EnvMomentoTokenProvider, configuration: Configuration, default_ttl_seconds: int
+):
+    async with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:172]: same

File: tests/test_momento_async.py [hunk 25]
```diff
@@ -293,43 +294,46 @@ async def test_set_with_null_value_throws_exception(client_async: SimpleCacheCli
...
 
 async def test_set_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, cache_name: str, default_ttl_seconds: int
+    bad_token_credential_provider: EnvMomentoTokenProvider,
+    configuration: Configuration,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    async with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client_async:
+    async with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:324]: same

File: tests/test_momento_async.py [hunk 26]
```diff
@@ -293,43 +294,46 @@ async def test_set_with_null_value_throws_exception(client_async: SimpleCacheCli
...
 async def test_set_throws_timeout_error_for_short_request_timeout(
-    auth_token: str, cache_name: str, default_ttl_seconds: int
+    configuration: Configuration,
+    credential_provider: EnvMomentoTokenProvider,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    async with SimpleCacheClient(auth_token, default_ttl_seconds, request_timeout_ms=1) as client_async:
+    configuration = configuration.with_client_timeout(timedelta(milliseconds=1))
+    async with SimpleCacheClient(configuration, credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:336]: same

File: tests/test_momento_async.py [hunk 27]
```diff
@@ -367,29 +369,34 @@ async def test_get_with_null_key_throws_exception(client_async: SimpleCacheClien
...
 
 async def test_get_throws_authentication_exception_for_bad_token(
-    bad_auth_token: str, cache_name: str, default_ttl_seconds: int
+    bad_token_credential_provider: EnvMomentoTokenProvider,
+    configuration: Configuration,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    async with SimpleCacheClient(bad_auth_token, default_ttl_seconds) as client_async:
+    async with SimpleCacheClient(configuration, bad_token_credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:387]: same

File: tests/test_momento_async.py [hunk 28]
```diff
@@ -367,29 +369,34 @@ async def test_get_with_null_key_throws_exception(client_async: SimpleCacheClien
...
 async def test_get_throws_timeout_error_for_short_request_timeout(
-    auth_token: str, cache_name: str, default_ttl_seconds: int
+    configuration: Configuration,
+    credential_provider: EnvMomentoTokenProvider,
+    cache_name: str,
+    default_ttl_seconds: int,
 ):
-    async with SimpleCacheClient(auth_token, default_ttl_seconds, request_timeout_ms=1) as client_async:
+    configuration = configuration.with_client_timeout(timedelta(milliseconds=1))
+    async with SimpleCacheClient(configuration, credential_provider, default_ttl_seconds) as client_async:
```
Comment (schwern) [inline tests/test_momento_async.py:399]: same

**Review by schwern — CHANGES_REQUESTED** [review]

File: tests/test_momento.py (hunk 10 above)
Comment (pgautier404) [reply tests/test_momento.py:54]: Good catch-- this error message wasn't accurate. The tests passed because the mistyped argument to `with_client_timeout()` also triggered the validation failure. Also python doesn't enforce typehints. The fixture being passed in as `default_ttl_seconds` actually is a timedelta, so was fine to work with the tests despite the incorrect type hint in the test method signature.

I updated the message to specify that the timeout needs to be a timedelta, left this test as it is because it's still useful, and added a new test with the negative value correctly expressed as a timedelta. I also corrected the other calls to `with_client_timeout()` that were passing integers and corrected the typehints for tests using the `default_ttl_seconds` fixture from int to timedelta.

File: src/momento/_utilities/_data_validation.py (hunk 1 above)
Comment (pgautier404) [reply src/momento/_utilities/_data_validation.py:32]: Good observation. I pushed the conversion from timedelta to seconds into the `prepare_set_request()` methods.

File: src/momento/aio/_scs_data_client.py (hunk 2 above)
Comment (pgautier404) [reply src/momento/aio/_scs_data_client.py:46]: Added a utility method to DRY this calculation out, as it is used by both sync and async clients.

File: src/momento/simple_cache_client.py (hunk 9 above)
Comment (pgautier404) [reply src/momento/simple_cache_client.py:37]: Not a bad idea. I added validation in `StaticGrpcConfiguration`'s `withClientTimeout()`. I'm going to leave this here as a defensive measure though.

File: tests/test_momento.py (hunk 10 above)
Comment (schwern) [reply tests/test_momento.py:54]: I wonder if we can get the tests to enforce type hints?

Comment (malandis) [reply tests/test_momento.py:66]: @schwern we're going to do that in short order; we don't run the type checker on the tests right now.

File: src/momento/aio/_scs_data_client.py [hunk 29]
```diff
@@ -42,13 +41,11 @@ def __init__(self, configuration: Configuration, credential_provider: Credential
         endpoint = credential_provider.get_cache_endpoint()
         self._logger = logs.logger
         self._logger.debug("Simple cache data client instantiated with endpoint: %s", endpoint)
-        default_deadline: timedelta = (
-            configuration.get_transport_strategy().get_grpc_configuration().get_deadline() or _DEFAULT_DEADLINE_SECONDS
-        )
+        default_deadline: timedelta = get_default_client_deadline(configuration)
```
Comment (schwern) [inline src/momento/aio/_scs_data_client.py:44]: I was thinking this could be a method on TransportStrategy, no function wrapper. It can also hide that we're using the GRPC deadline as a timeout.

TransportStrategy defines an abstract `get_client_timeout` and StaticTransportStrategy implements it. The default lives in StaticTransportStrategy.

I think we should avoid the GRPC term "deadline" when the transport strategy is hiding GRPC. We even [pass the "deadline" as a "timeout" to the GRPC stub](https://github.com/momentohq/client-sdk-python/blob/51ecff677cee47ab7d0012838399f3bcaf5ec454/src/momento/aio/_scs_data_client.py#L65).

```
default_timeout: configuration.get_transport_strategy.get_client_timeout()
```

File: tests/test_momento.py (hunk 10 above)
Comment (schwern) [reply tests/test_momento.py:66]: @pgautier404 I've run into that issue in Ruby, too. I've gotten around it by defining subclasses of InvalidArgumentError and checking for specific ones.

**Review by schwern — APPROVED** [review]
lgtm. Up to you if you want to make the suggested changes.

**Review by schwern — APPROVED** [review]
